src/onegov/core/types.py

```diff
@@ -56,5 +68,99 @@ class LaxFileDict(TypedDict):
     _T = TypeVar('_T')
     SequenceOrScalar = Union[Sequence[_T], _T]
 
-    class PaginatedGenericCollection(GenericCollection[_M], Pagination[_M]):
-        pass
+    class PaginatedGenericCollection(Protocol[_M]):
+        """ Intersection type of GenericCollection and Pagination, as
+          implemented by for example:
+          PaginatedAgencyCollection(GenericCollection, Pagination)"""
+
+        # GenericCollection
+        @property
+        def model_class(self) -> type[_M]:
+            ...
+
+        @cached_property
+        def primary_key(self) -> 'Column[str] | Column[UUID] | Column[int]':
+            ...
+
+        def query(self) -> 'Query[_M]':
+            ...
+
+        def by_id(self, id: 'PKType') -> _M | None:
+            ...
+
+        def by_ids(self, ids: 'Collection[PKType]') -> list[_M]:
+            ...
+
+        def add(self, **kwargs: Any) -> _M:
+            ...
+
+        def add_by_form(
+            self,
+            form: '_FormThatSupportsGetUsefulData',
+            properties: 'Iterable[str] | None' = None,
+        ) -> _M:
+            ...
+
+        def delete(self, item: _M) -> None:
+            ...
+
+        # Pagination:
+        batch_size: int
+
+        def __eq__(self, other: object) -> bool:
+            ...
+
+        def subset(self) -> 'Query[_M]':
+            ...
+
+        @cached_property
+        def cached_subset(self) -> 'Query[_M]':
+            ...
+
+        @property
+        @abstractmethod
```

Comment:
`abstractmethod` is redundant inside a protocol, mypy will probably get confused about this when checking the protocol, so it's better to get rid of it.

Also not everyone seems to use `page` consistently, I have seen some collections where `page` is optional. So it might be better to annotate it as `page: int | None`, so it matches as many implementations as possible.